File: crates/netcfgd-sys/src/tun.rs
//! Persistent tun and tap devices, which netlink cannot create.
//!
//! **The one link kind that is not an `RTM_NEWLINK`.** Every other virtual link
//! netcfgd makes -- a bridge, a bond, a VLAN, a VXLAN, a tunnel, a veth pair --
//! is a netlink message with a kind name in it. A tun or tap device is not: it
//! comes from a `TUNSETIFF` ioctl on `/dev/net/tun`, and it exists only as long
//! as something holds that descriptor unless `TUNSETPERSIST` is set on it.
//!
//! That is why the model carried [`crate::ops::NewLink`]'s cousin for years
//! with "in the schema and not implemented" written on it: an ioctl is
//! `unsafe`, and `unsafe` lives in this crate alone (constraint 4). So it
//! belongs here, beside the terminal size and `SO_PEERCRED`, rather than being
//! a permanent hole in what the configuration language can ask for.
//!
//! **What `ip tuntap add` does, in the same order.** Open the clone device, ask
//! for a name and a mode, hand it an owner and a group where the configuration
//! names them, then make it persist and close. The device outlives this process
//! because persistence is what the last ioctl asks for; closing without it
//! would delete the device again, which is the failure mode worth knowing about
//! before reading the code.

use std::io;
use std::os::fd::{AsRawFd as _, OwnedFd};

/// `TUNSETIFF`, `TUNSETPERSIST`, `TUNSETOWNER`, `TUNSETGROUP`.
///
/// Written out rather than taken from `libc`, which does not export them: each
/// is `_IOW('T', n, int)` and the encoding is stable kernel ABI. The numbers
/// are checked against `linux/if_tun.h` and, more usefully, against `ip tuntap`
/// producing the same device -- which is what the live test compares.
const TUNSETIFF: libc::c_ulong = 0x4004_54ca;
const TUNSETPERSIST: libc::c_ulong = 0x4004_54cb;
const TUNSETOWNER: libc::c_ulong = 0x4004_54cc;
const TUNSETGROUP: libc::c_ulong = 0x4004_54ce;

/// The clone device every tun and tap comes from.
const CLONE_DEVICE: &str = "/dev/net/tun";

/// `struct ifreq`, as `TUNSETIFF` reads it.
///
/// Declared here rather than borrowed from `libc::ifreq`, whose union arm
/// depends on the libc version and which would make the one field this needs --
/// the flags -- reachable only through a second `unsafe` block. The kernel
/// reads `ifr_name` and `ifr_flags` and nothing else from this call, and the
/// tail is padding it ignores.
#[repr(C)]
struct IfReq {
	name: [libc::c_char; libc::IFNAMSIZ],
	flags: libc::c_short,
	padding: [u8; 22],
}

/// Whether the device carries IP packets or ethernet frames.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Mode {
	/// Layer 3: IP packets, no ethernet header.
	Tun,
	/// Layer 2: full ethernet frames.
	Tap,
}

impl Mode {
	/// The flag `TUNSETIFF` wants, with `IFF_NO_PI` beside it.
	///
	/// **`IFF_NO_PI` always**, which is what `ip tuntap add` does without
	/// saying so. Without it every packet carries a four-byte protocol header
	/// that nothing else on the machine expects, and the device looks subtly
	/// broken to whatever attaches to it rather than failing to appear.
	fn flags(self) -> libc::c_short {
		let kind = match self {
			Self::Tun => libc::IFF_TUN,
			Self::Tap => libc::IFF_TAP,
		};
		#[allow(clippy::cast_possible_truncation)]
		{
			(kind | libc::IFF_NO_PI) as libc::c_short
		}
	}
}

/// Make a persistent tun or tap device.
///
/// `owner` and `group` are the ids permitted to attach to it, where the
/// configuration names them. A device with neither may be attached to by root
/// alone, which is the kernel's default and not something this invents.
///
/// # Errors
///
/// The errno from whichever step failed, with the step named: `EBUSY` from the
/// first ioctl means a device by that name already exists and is not a tun,
/// `EPERM` means no `CAP_NET_ADMIN`, and `ENOENT` on the open means the `tun`
/// module is not loaded -- three different things to do about it, so they are
/// not flattened into one message.
pub fn create(name: &str, mode: Mode, owner: Option<u32>, group: Option<u32>) -> io::Result<()> {
	if name.is_empty() || name.len() >= libc::IFNAMSIZ {
		return Err(io::Error::new(
			io::ErrorKind::InvalidInput,
			format!("`{name}` is not a name the kernel would take for a link"),
		));
	}

	let file = std::fs::OpenOptions::new()
		.read(true)
		.write(true)
		.open(CLONE_DEVICE)
		.map_err(|error| {
			io::Error::new(
				error.kind(),
				format!("cannot open {CLONE_DEVICE}: {error}; is the `tun` module loaded?"),
			)
		})?;
	let fd = OwnedFd::from(file);

	let mut request = IfReq {
		name: [0; libc::IFNAMSIZ],
		flags: mode.flags(),
		padding: [0; 22],
	};
	for (slot, byte) in request.name.iter_mut().zip(name.as_bytes()) {
		#[allow(clippy::cast_possible_wrap)]
		{
			*slot = *byte as libc::c_char;
		}
	}

	// SAFETY: `TUNSETIFF` takes exactly one pointer to a `struct ifreq`, which
	// is what is passed and which is live for the duration of the call. The
	// name is NUL-terminated by construction -- the array is zeroed and the
	// copy above is refused at `IFNAMSIZ - 1` bytes -- and the kernel writes
	// back at most the name it chose, which this does not read.
	let created =
		unsafe { libc::ioctl(fd.as_raw_fd(), TUNSETIFF, std::ptr::addr_of_mut!(request)) };
	if created < 0 {
		let error = io::Error::last_os_error();
		return Err(io::Error::new(
			error.kind(),
			format!("TUNSETIFF for {name}: {error}"),
		));
	}

	// Before persistence, so a device that cannot be given its owner is not
	// left behind for somebody to find: closing the descriptor here removes it.
	for (what, value, request) in [
		("TUNSETOWNER", owner, TUNSETOWNER),
		("TUNSETGROUP", group, TUNSETGROUP),
	] {
		let Some(value) = value else {
			continue;
		};
		// SAFETY: both take an integer by value rather than a pointer, which is
		// what the `_IOW('T', n, int)` encoding in the constants above says.
		let set = unsafe { libc::ioctl(fd.as_raw_fd(), request, libc::c_ulong::from(value)) };
		if set < 0 {
			let error = io::Error::last_os_error();
			return Err(io::Error::new(
				error.kind(),
				format!("{what} for {name}: {error}"),
			));
		}
	}

	// **Last, and the reason the device survives this function.** A tun device
	// exists only while something holds its descriptor; without this the close
	// below would delete what was just made, and the call would report success
	// over a machine with no such link.
	//
	// SAFETY: takes an integer by value, as above.
	let persisted = unsafe { libc::ioctl(fd.as_raw_fd(), TUNSETPERSIST, 1_u64) };
	if persisted < 0 {
		let error = io::Error::last_os_error();
		return Err(io::Error::new(
			error.kind(),
			format!("TUNSETPERSIST for {name}: {error}"),
		));
	}
	Ok(())
}

File: c/src/sys/tun.c
/*
 * tun.c -- the four ioctls that make a persistent tun or tap.
 *
 * The whole of the ordering argument is in tun.h and in decision 0254. What is
 * here is the two halves that argument needs kept apart: `ncfg_tun_plan`, which
 * decides what to do and touches nothing, and `ncfg_tun_create`, which opens
 * one descriptor and performs the plan without deciding anything.
 *
 * That split is the same one `wire.c` and `socket.c` have, and it is what lets
 * the order be asserted on a machine that must not have a device made on it.
 */
#include "ncfg/tun.h"

#include <errno.h>
#include <fcntl.h>
#include <stddef.h>
#include <string.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include <linux/if_tun.h>
#include <net/if.h>

/*
 * The request this module builds is the libc's `struct ifreq`, and these are
 * what keep that claim honest. `net/if.h` is included for exactly this -- not
 * for the type, which tun.h declares for the reason given there.
 */
_Static_assert(sizeof(struct ifreq) == NCFG_TUN_REQUEST_LEN,
    "a tun request is forty bytes and this libc's ifreq is not");
_Static_assert(sizeof(ncfg_tun_request_t) == NCFG_TUN_REQUEST_LEN,
    "and the one built here is not forty bytes");
_Static_assert(offsetof(ncfg_tun_request_t, flags) == NCFG_TUN_NAME_MAX,
    "the flags must sit where the kernel reads ifr_flags");
_Static_assert(offsetof(struct ifreq, ifr_flags) == NCFG_TUN_NAME_MAX,
    "and that is not where this libc keeps them");
_Static_assert(IFNAMSIZ == NCFG_TUN_NAME_MAX,
    "an interface name is sixteen bytes and this libc says otherwise");

const char *ncfg_tun_mode_name(ncfg_tun_mode_t mode)
{
	switch (mode) {
	case NCFG_TUN_MODE_TUN:
		return "tun";
	case NCFG_TUN_MODE_TAP:
		return "tap";
	}
	return NULL;
}

int ncfg_tun_mode_from_name(const char *text, ncfg_tun_mode_t *out)
{
	if (!text || !out) {
		return 0;
	}
	if (strcmp(text, "tun") == 0) {
		*out = NCFG_TUN_MODE_TUN;
		return 1;
	}
	if (strcmp(text, "tap") == 0) {
		*out = NCFG_TUN_MODE_TAP;
		return 1;
	}
	return 0;
}

int16_t ncfg_tun_flags(ncfg_tun_mode_t mode)
{
	switch (mode) {
	case NCFG_TUN_MODE_TUN:
		return (int16_t)(IFF_TUN | IFF_NO_PI);
	case NCFG_TUN_MODE_TAP:
		return (int16_t)(IFF_TAP | IFF_NO_PI);
	}
	return 0;
}

int ncfg_tun_request(const ncfg_tun_spec_t *spec, ncfg_tun_request_t *out,
    char *err, size_t err_size)
{
	size_t length;

	if (!out) {
		ncfg_error_set(err, err_size, "a tun request needs somewhere to go");
		return 0;
	}
	memset(out, 0, sizeof(*out));
	if (!spec || !spec->name) {
		ncfg_error_set(err, err_size, "making a tun device needs a name");
		return 0;
	}
	length = strlen(spec->name);
	/* `>=`, not `>`: the kernel's field is sixteen bytes *including* the
	 * terminator, and a name that filled it would come back as whatever the
	 * next field happens to hold. */
	if (length == 0 || length >= NCFG_TUN_NAME_MAX) {
		ncfg_error_set(err, err_size,
		    "`%s` is not a name the kernel would take for a link", spec->name);
		return 0;
	}
	out->flags = ncfg_tun_flags(spec->mode);
	if (out->flags == 0) {
		ncfg_error_set(err, err_size, "a tun device is either a tun or a tap");
		return 0;
	}
	/* Into a zeroed array, so the name is NUL-terminated by construction
	 * rather than by a terminator somebody remembered to write. */
	memcpy(out->name, spec->name, length);
	return 1;
}

int ncfg_tun_plan(const ncfg_tun_spec_t *spec, ncfg_tun_step_t *out, size_t max,
    size_t *count, char *err, size_t err_size)
{
	ncfg_tun_request_t request;
	size_t             at = 0;

	if (count) {
		*count = 0;
	}
	if (!out || !count || max < NCFG_TUN_STEPS_MAX) {
		ncfg_error_set(err, err_size, "a tun plan needs room for %u steps",
		    (unsigned)NCFG_TUN_STEPS_MAX);
		return 0;
	}
	memset(out, 0, max * sizeof(*out));
	/* Built and thrown away: what is wanted here is the refusal for a name
	 * the kernel would not take, before a descriptor is opened for it. */
	if (!ncfg_tun_request(spec, &request, err, err_size)) {
		return 0;
	}

	out[at].what = "TUNSETIFF";
	out[at].request = TUNSETIFF;
	out[at].is_request_struct = 1;
	at++;

	/* Before persistence, so a device that cannot be given its owner is not
	 * left behind for somebody to find: closing the descriptor at that point
	 * removes it. See tun.h. */
	if (spec->has_owner) {
		out[at].what = "TUNSETOWNER";
		out[at].request = TUNSETOWNER;
		out[at].value = (unsigned long)spec->owner;
		at++;
	}
	if (spec->has_group) {
		out[at].what = "TUNSETGROUP";
		out[at].request = TUNSETGROUP;
		out[at].value = (unsigned long)spec->group;
		at++;
	}

	/* **Last, and the reason the device survives this call.** A tun device
	 * exists only while something holds its descriptor; without this the
	 * close would delete what was just made, and the call would report
	 * success over a machine with no such link. */
	out[at].what = "TUNSETPERSIST";
	out[at].request = TUNSETPERSIST;
	out[at].value = 1u;
	at++;

	*count = at;
	return 1;
}

int ncfg_tun_create(const char *clone_device, const ncfg_tun_spec_t *spec,
    char *err, size_t err_size)
{
	ncfg_tun_request_t request;
	ncfg_tun_step_t    steps[NCFG_TUN_STEPS_MAX];
	size_t             count = 0;
	size_t             at;
	int                fd;

	if (!clone_device) {
		ncfg_error_set(err, err_size, "making a tun device needs the clone device");
		return 0;
	}
	/* The name is refused here, before anything is opened: a refusal that
	 * arrives after the open is a refusal that had a descriptor to leak. */
	if (!ncfg_tun_request(spec, &request, err, err_size)) {
		return 0;
	}
	if (!ncfg_tun_plan(spec, steps, NCFG_TUN_STEPS_MAX, &count, err, err_size)) {
		return 0;
	}

	/* Read and write, because `TUNSETIFF` wants a writable descriptor, and
	 * `O_CLOEXEC` because netcfgd spawns children and a descriptor on this
	 * device outliving the call would keep an unpersisted device alive in
	 * something that knows nothing about it. */
	fd = open(clone_device, O_RDWR | O_CLOEXEC);
	if (fd < 0) {
		ncfg_error_set(err, err_size, "cannot open %s: %s%s", clone_device,
		    strerror(errno),
		    errno == ENOENT ? "; is the `tun` module loaded?" : "");
		return 0;
	}
	for (at = 0; at < count; at++) {
		int done;

		if (steps[at].is_request_struct) {
			done = ioctl(fd, steps[at].request, &request);
		} else {
			done = ioctl(fd, steps[at].request, steps[at].value);
		}
		if (done < 0) {
			int failed = errno;

			/* Closed before the message is built, and this is the
			 * close that removes a half-made device: nothing has
			 * persisted yet unless the last step succeeded, and the
			 * last step is the only one after which this loop does
			 * not run again. */
			(void)close(fd);
			ncfg_error_set(err, err_size, "%s for %s: %s", steps[at].what,
			    spec->name, strerror(failed));
			return 0;
		}
	}
	(void)close(fd);
	return 1;
}
